File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.cpp
#include<string>
#include <iostream>
#include <stack>
using namespace std;

bool solution(string s)
{
    bool answer = true;
    stack<char> st;
    for(auto x:s){
        if(x=='('){
            st.push(x);
        }
        else{
            if(st.empty()){
                answer = false;
                break;
            }
            else{
                st.pop();
            }
        }
    }
    if(!st.empty()){
        answer = false;
    }

    return answer;
}

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.java
import java.util.*;
class Solution {
    boolean solution(String s) {
        boolean answer = true;
        Deque<Character> dq = new ArrayDeque<>();
        for(int i=0;i<s.length();i++){
            char c = s.charAt(i);
            if(c == '('){
                dq.push(c);
            }
            else{
                if(dq.isEmpty()){
                    answer = false;
                    break;
                }
                char tmp = dq.peek();
                dq.pop();
                if(tmp!='('){
                    answer = false;
                    break;
                }
            }
        }
        if(!dq.isEmpty()){
            answer = false;
        }
        return answer;
    }
}